File: ts-sdk/src/findings/finding.ts
import { EntityType, Label } from "../labels";
import { assertIsFromEnum, assertIsNonEmptyString } from "../utils";
import { FindingSeverity } from "./finding.severity";
import { FindingSource } from "./finding.source";
import { FindingType } from "./finding.type";

type FindingInput = {
  name: string;
  description: string;
  alertId: string;
  protocol?: string;
  severity: FindingSeverity;
  type: FindingType;
  metadata?: { [key: string]: string };
  addresses?: string[];
  labels?: Label[];
  uniqueKey?: string;
  source?: FindingSource;
  timestamp?: Date;
};

export class Finding {
  private constructor(
    readonly name: string,
    readonly description: string,
    readonly alertId: string,
    readonly protocol: string,
    readonly severity: FindingSeverity,
    readonly type: FindingType,
    readonly metadata: { [key: string]: string },
    readonly addresses: string[],
    readonly labels: Label[],
    readonly uniqueKey: string,
    readonly source: FindingSource,
    readonly timestamp: Date
  ) {}

  toString() {
    return JSON.stringify(
      {
        ...this,
        severity: FindingSeverity[this.severity],
        type: FindingType[this.type],
        labels: this.labels.map((l) =>
          Object.assign(l, {
            entityType: EntityType[l.entityType],
          })
        ),
      },
      null,
      2
    );
  }

  static from(findingInput: FindingInput) {
    return this.fromObject(findingInput);
  }

  static fromObject({
    name,
    description,
    alertId,
    severity,
    type,
    timestamp = new Date(Math.floor(Date.now() / 1000) * 1000),
    metadata = {},
    addresses = [],
    labels = [],
    uniqueKey = "",
    source = {},
    protocol = "",
  }: FindingInput) {
    assertIsNonEmptyString(name, "name");
    assertIsNonEmptyString(description, "description");
    assertIsNonEmptyString(alertId, "alertId");
    assertIsFromEnum(severity, FindingSeverity, "severity");
    assertIsFromEnum(type, FindingType, "type");
    // TODO assert metadata keys and values are strings

    labels = labels.map((l) => (l instanceof Label ? l : Label.fromObject(l)));

    return new Finding(
      name,
      description,
      alertId,
      protocol,
      severity,
      type,
      metadata,
      addresses,
      labels,
      uniqueKey,
      source,
      timestamp
    );
  }
}

File: py-sdk/src/forta_bot_sdk/findings/finding.py
import time
from datetime import datetime
from typing import Optional, TypedDict
from ..utils import assert_is_non_empty_string, assert_is_from_enum, assert_is_string_key_to_string_value_map, JSONable
from ..labels import Label
from .finding_severity import FindingSeverity
from .finding_type import FindingType
from .finding_source import FindingSource


class FindingInput(TypedDict):
    name: str
    description: str
    alert_id: str
    severity: FindingSeverity
    type: FindingType
    protocol: Optional[str]
    metadata: Optional[dict]
    addresses: Optional[list[str]]
    labels: Optional[list[Label]]
    unique_key: Optional[str]
    source: Optional[FindingSource]
    timestamp: datetime


class Finding(JSONable):
    def __init__(self, dict: FindingInput):
        self.name: str = assert_is_non_empty_string(dict.get('name'), 'name')
        self.description: str = assert_is_non_empty_string(
            dict.get('description'), 'description')
        self.alert_id: str = assert_is_non_empty_string(
            dict.get('alert_id'), 'alert_id')
        self.severity: FindingSeverity = assert_is_from_enum(
            dict.get('severity'), FindingSeverity, 'severity')
        self.type: FindingType = assert_is_from_enum(
            dict.get('type'), FindingType, 'type')
        self.metadata: dict[str, str] = assert_is_string_key_to_string_value_map(
            dict.get('metadata'), 'metadata')
        self.addresses: list[str] = dict.get('addresses')
        self.labels: list[Label] = [Label(l) if not isinstance(
            l, Label) else l for l in dict.get('labels', [])]
        self.unique_key: str = dict.get('unique_key')
        self.source: FindingSource = FindingSource(
            dict.get('source')) if dict.get('source') is not None else None
        self.timestamp: datetime = dict.get(
            'timestamp', datetime.fromtimestamp(int(time.time())))
